### PR TITLE
[Jax] Implement Fused QKV Projections to Reduce HBM-to-VMEM Redundant Data Loading

### PR DESCRIPTION
**Pull request by JiriesKaileh**
# Description

Issue
Currently, for Gemma 4, Q, K, and V projections are handled as separate operations. This requires the model to load the same input data from HBM to VMEM three separate times; once for each projection.

Proposed Solution
To get past this issue, we will fuse the Q, K, and V projections into a single operation. By merging these, the input data is loaded from HBM into VMEM once, and all weight sums for Q, K, and V are performed simultaneously

High level implementation details
Create a flax module and delegate to the existing kernel. There is also some [preprocessing details](https://b.corp.google.com/issues/467153577#comment10) that are needed to eliminate the all-to-all collective.

Results: 
5x improvement in per-layer performance

See more in this [bug]( https://b.corp.google.com/issues/520442813)

# Tests

- `google/gemma-4-31B-it` boots healthy with `MODEL_IMPL_TYPE=flax_nnx`

```
SKIP_JAX_PRECOMPILE=1 USE_BATCHED_RPA_KERNEL=1 MODEL_IMPL_TYPE=flax_nnx vllm \
    serve google/gemma-4-31B-it \
    --max-model-len=1524 \
    --max-num-seqs=256 \
    --data-parallel-size 4 \
    --tensor-parallel-size 2 \
    --max-num-batched-tokens 16384 \
    --no-enable-prefix-caching \
    --additional_config='{"quantization": { "qwix": { "rules": [{ "module_path": ".*", "weight_qtype": "float8_e4m3fn", "act_qtype": "float8_e4m3fn"}]}}}' \
    --kv-cache-dtype=fp8 \
    --gpu-memory-utilization=0.9 \
    --async-scheduling \
    --limit-mm-per-prompt '{"image": 0, "video": 0, "audio": 0}' \
    --block-size=256
```
Accuracy is [0.974](https://paste.googleplex.com/4788992364118016): `{'accuracy': 0.9714, 'gen_num': 140}`

```
python scripts/vllm/benchmarking/benchmark_serving.py \
      --backend vllm \
      --model google/gemma-4-31B-it \
      --dataset-name mmlu \
      --dataset-path /mnt/disks/jiries-disk_data/mmlu/data/test \
      --mmlu-use-chat-template \
      --chat-template-system-prompt "Reasoning effort: high, skip thinking process and must only give short answer in something
  like A, B, C, D." \
      --mmlu-output-len=16 \
      --run-eval \
      --num-prompts 140
```

### CODE REVIEW

File: tpu_inference/layers/vllm/custom_ops/linear.py [hunk 1]
```diff
@@ -158,69 +153,16 @@ def forward(
...
 
-        out_jax = reorder_concatenated_tensor_for_sharding(
-            out_jax,
-            self.output_sizes,
-            self.tp_size,
-            dim=-1,
-        )
-        q_jax, k_jax, v_jax = slice_sharded_tensor_for_concatenation(
+        from tpu_inference.layers.common.utils import \
+            process_sharded_qkv_projection
```
Comment (lk-chen) [inline tpu_inference/layers/vllm/custom_ops/linear.py]: let's rename to `slice_qkv_merge_projection_output` and move the import to top

File: tpu_inference/layers/common/utils.py [hunk 2]
```diff
@@ -178,3 +181,123 @@ def cpu_mesh() -> Mesh:
...
     """A context to enforce using CPU mesh, used for loading weights on CPU."""
     return jax.set_mesh(cpu_mesh())
+
+
+def process_sharded_qkv_projection(
+    out_jax: jax.Array,
+    output_sizes: list[int],
+    tp_size: int,
+    num_kv_head_replicas: int = 1,
+    reorder_output: bool = True,
```
Comment (lk-chen) [inline tpu_inference/layers/common/utils.py]: I don't see it set to False in this PR? also the next param

File: tpu_inference/layers/vllm/custom_ops/linear.py [hunk 3]
```diff
@@ -158,69 +153,16 @@ def forward(
...
             self.output_sizes,
             self.tp_size,
+            num_kv_head_replicas=self.num_kv_head_replicas,
         )
 
-        mesh = jax.sharding.get_abstract_mesh()
-        # Split the `kv_head` mesh axis (size kv_size) into a pair
-        # `(kv_head, replica)` of sizes `(kv_size // replicas, replicas)`.
-        # The `replica` sub-axis is later replicated via `shard_map` with no
-        # data movement, since `_tile_kv` already placed identical KV-head
```
Comment (lk-chen) [inline tpu_inference/layers/vllm/custom_ops/linear.py:177]: let's keep these comments in new code

File: tpu_inference/layers/jax/linear.py [hunk 4]
```diff
@@ -169,3 +170,74 @@ def __init__(self,
...
+            bias_init=nnx.with_partitioning(nnx.initializers.uniform(),
+                                            ("model", )) if use_bias else None,
+            rngs=rngs,
+            quant_config=quant_config,
+            prefix=prefix + ".qkv_proj",
+        )
+
+    def __call__(self, x: jax.Array) -> Tuple[jax.Array, jax.Array, jax.Array]:
+        from tpu_inference.layers.common.utils import \
+            slice_sharded_tensor_for_causal_lm
```
Comment (lk-chen) [inline tpu_inference/layers/jax/linear.py]: move to top if possible?

File: tpu_inference/layers/jax/linear.py [hunk 5]
```diff
@@ -169,3 +170,74 @@ def __init__(self,
...
+
+class JaxQKVParallelLinear(JaxModule):
+    """Fused QKV Parallel Linear layer for JAX-native models.
+
+    Performs fused Q, K, and V projections in a single HBM read pass and
+    partitions them locally per TPU device without incurring all-to-all collectives.
+    """
+
+    def __init__(self,
+                 hidden_size: int,
```
Comment (lk-chen) [inline tpu_inference/layers/jax/linear.py:232]: put "*" after "self" to force use kwargs

File: tpu_inference/layers/jax/linear.py [hunk 6]
```diff
@@ -169,3 +170,74 @@ def __init__(self,
...
+        self.tp_size = tp_size
+
+        self.q_size = num_heads * head_dim
+        self.k_size = num_kv_heads * head_dim
+        self.v_size = num_kv_heads * head_dim
+        self.total_output_dim = self.q_size + self.k_size + self.v_size
+
+        # Output dimension is sharded along the "model" (TP) axis
+        self.proj = JaxEinsum(
+            "TD,DV->TV",
```
Comment (lk-chen) [inline tpu_inference/layers/jax/linear.py]: where does the einsum str come from? usually we put meaningful string here, "D" for model dim, "V" for vocab size. Here you should use "TD,DA->TA", A for "all dim"

File: tpu_inference/layers/jax/linear.py [hunk 7]
```diff
@@ -169,3 +170,74 @@ def __init__(self,
...
+        # with zero abstract-mesh tracking or shard_map calls. This ensures JAX traces and inlines
+        # it as a pure metadata-only zero-copy slice and reshape with 100% optimal HLO speed.
+        outs_q, outs_k, outs_v = slice_sharded_tensor_for_causal_lm(
+            outs, self.q_size, self.k_size, self.v_size, self.tp_size)
+
+        # Reshape directly to their global multi-head shapes (metadata change under JAX!)
+        q = outs_q.reshape(outs.shape[:-1] + (self.num_heads, self.head_dim))
+        k = outs_k.reshape(outs.shape[:-1] +
+                           (self.num_kv_heads, self.head_dim))
+        v = outs_v.reshape(outs.shape[:-1] +
```
Comment (lk-chen) [inline tpu_inference/layers/jax/linear.py]: does it handle the case v is optional? or you plan to finish in next PR (leave TODO please)?

File: tpu_inference/models/jax/gemma4_mm.py [hunk 8]
```diff
@@ -645,40 +649,170 @@ def filter_weights(weights_iterator):
                         )):
                     continue
 
+                # Handle separate Q, K, V projections and fuse them on the fly if needed
+                is_separate_proj = False
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4_mm.py]: vLLM uses [packed_modules_mapping](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/gemma4.py#L1526-L1530) let's see if you can reuse https://github.com/vllm-project/tpu-inference/pull/2859

**Review by lk-chen — CHANGES_REQUESTED** [review]

File: tpu_inference/layers/vllm/custom_ops/linear.py (hunk 1 above)
Comment (JiriesKaileh) [reply tpu_inference/layers/vllm/custom_ops/linear.py]: Done

File: tpu_inference/layers/common/utils.py (hunk 2 above)
Comment (JiriesKaileh) [reply tpu_inference/layers/common/utils.py]: Removed

File: tpu_inference/layers/vllm/custom_ops/linear.py (hunk 3 above)
Comment (JiriesKaileh) [reply tpu_inference/layers/vllm/custom_ops/linear.py:177]: Reverted

File: tpu_inference/layers/jax/linear.py (hunk 4 above)
Comment (JiriesKaileh) [reply tpu_inference/layers/jax/linear.py]: Done

File: tpu_inference/layers/jax/linear.py (hunk 5 above)
Comment (JiriesKaileh) [reply tpu_inference/layers/jax/linear.py:232]: Done

File: tpu_inference/layers/jax/linear.py (hunk 6 above)
Comment (JiriesKaileh) [reply tpu_inference/layers/jax/linear.py]: Done

File: tpu_inference/layers/jax/linear.py (hunk 7 above)
Comment (JiriesKaileh) [reply tpu_inference/layers/jax/linear.py]: I'll address this in the next PR. I put a TODO

File: tpu_inference/models/jax/gemma4_mm.py (hunk 8 above)
Comment (JiriesKaileh) [reply tpu_inference/models/jax/gemma4_mm.py]: Done

File: tpu_inference/models/jax/gemma4.py [hunk 9]
```diff
@@ -452,13 +455,16 @@ def __call__(
...
-        if self.v_proj is None:
-            v = k
+        if self.qkv_proj is not None:
+            q, k, v = self.qkv_proj(x)
         else:
-            v = self.v_proj(x)
-        # q: (T, N, H)
-        q = self.q_proj(x)
+            k = self.k_proj(x)
+            if self.v_proj is None:
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4.py]: it should always be None? (line 368 in this file

Comment (JiriesKaileh) [reply tpu_inference/models/jax/gemma4.py]: Yes, fixed

File: tpu_inference/layers/jax/linear.py [hunk 10]
```diff
@@ -171,6 +174,156 @@ def __call__(self, inputs: jax.Array) -> jax.Array:
...
+        self.use_bias = use_bias
+        self.tp_size = tp_size
+
+        self.q_size = num_heads * head_dim
+        self.k_size = num_kv_heads * head_dim
+        self.v_size = num_kv_heads * head_dim
+        self.total_output_dim = self.q_size + self.k_size + self.v_size
+
+        # Output dimension is sharded along the "model" (TP) axis
+        self.proj = JaxEinsum(
```
Comment (lk-chen) [inline tpu_inference/layers/jax/linear.py]: Please refer what vLLM does https://github.com/vllm-project/vllm/blob/e0871ad2259768add6dc43e2972bd364d0d13086/vllm/model_executor/layers/linear.py#L1038-L1042

JaxQKVParallelLinear could inherit from JaxMergedColumnParallelLinear and reuse lots of logic there and pass [q_sz, k_sz, v_sz] to output_sizes

File: tpu_inference/models/jax/gemma4.py [hunk 11]
```diff
@@ -1077,8 +1090,181 @@ def load_weights(self, weights: Iterable[Tuple[str, Any]]):
...
+            import re
+
+            import torch
+
+            separate_cache = {}  # (layer_idx, suffix) -> {proj_type: weight}
+
+            for name, weight in weights_iterator:
+                # Handle separate Q, K, V projections and fuse them on the fly if needed
+                is_separate_proj = False
+                for proj_type in ("q_proj", "k_proj", "v_proj"):
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4.py]: prefer not having these kind of logic in each model, should already handled in layer (e.g. JaxMergedColumnParallelLinear)

File: tpu_inference/models/jax/gemma4.py [hunk 12]
```diff
@@ -1077,8 +1090,181 @@ def load_weights(self, weights: Iterable[Tuple[str, Any]]):
...
+                                            rearranged, dim=0)
+
+                                        target_name = name.replace(
+                                            proj_type, "qkv_proj.proj")
+                                        yield target_name, rearranged_tensor
+
+                if is_separate_proj:
+                    continue
+
+                if "qkv_proj" in name:
```
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4.py]: Same here, I think you need to special process the case where weight is already fused for JaxMergedColumnParallelLinear. vllm has these logic (https://github.com/vllm-project/vllm/blob/e0871ad2259768add6dc43e2972bd364d0d13086/vllm/model_executor/layers/linear.py#L1091-L1092) but a bit too complicated to borrow

File: tpu_inference/models/jax/gemma4.py (hunk 11 above)
Comment (JiriesKaileh) [reply tpu_inference/models/jax/gemma4.py]: Removed

File: tpu_inference/models/jax/gemma4.py (hunk 12 above)
Comment (JiriesKaileh) [reply tpu_inference/models/jax/gemma4.py]: Removed

File: tpu_inference/layers/common/utils.py [hunk 13]
```diff
@@ -178,3 +181,108 @@ def cpu_mesh() -> Mesh:
 def cpu_mesh_context():
     """A context to enforce using CPU mesh, used for loading weights on CPU."""
     return jax.set_mesh(cpu_mesh())
+
+
+def slice_qkv_merge_projection_output(
```
Comment (lk-chen) [inline tpu_inference/layers/common/utils.py]: not used in this PR

File: tests/layers/jax/test_linear.py [hunk 14]
```diff
@@ -163,6 +163,139 @@ def test_sharding_assignment(self):
         self.assertEqual(f"{jax.typeof(jax_linear.weight.value)}",
                          "float32[16,32]")
 
+    def test_jax_qkv_parallel_linear_correctness(self):
+        """Tests JaxQKVParallelLinear correctness against separate JaxEinsum projections."""
+        from jax import numpy as jnp
+
+        from tpu_inference.layers.common.utils import \
+            reorder_concatenated_tensor_for_sharding
+        from tpu_inference.layers.jax.linear import (JaxEinsum,
```
Comment (lk-chen) [inline tests/layers/jax/test_linear.py]: move import to top of file

File: tests/layers/jax/test_linear.py [hunk 15]
```diff
@@ -163,6 +163,139 @@ def test_sharding_assignment(self):
...
+        head_dim = 8
+        seq_len = 3
+        use_bias = True
+
+        num_devices = len(jax.devices('cpu'))
+        tp_sizes = [1]
+        if num_devices >= 2:
+            tp_sizes.append(2)
+
+        for tp_size in tp_sizes:
```
Comment (lk-chen) [inline tests/layers/jax/test_linear.py]: should use parameterized test instead of for loop

File: tpu_inference/layers/jax/quantization/unquantized.py [hunk 16]
```diff
@@ -117,7 +121,15 @@ def _load_merged_weight(param: nnx.Param, torch_weight, shard_id: int, *,
         once every slot is filled.
         """
         shards = param.get_metadata("_merged_shards")
-        shards[shard_id] = torch_weight
+        if shard_id == -1:  #TODO: Ensure this is the right logic for splitting
```
Comment (lk-chen) [inline tpu_inference/layers/jax/quantization/unquantized.py]: is it still TODO?

File: tpu_inference/layers/common/utils.py (hunk 13 above)
Comment (JiriesKaileh) [reply tpu_inference/layers/common/utils.py]: Removed

File: tests/layers/jax/test_linear.py (hunk 14 above)
Comment (JiriesKaileh) [reply tests/layers/jax/test_linear.py]: Done

File: tests/layers/jax/test_linear.py (hunk 15 above)
Comment (JiriesKaileh) [reply tests/layers/jax/test_linear.py]: Done

File: tpu_inference/layers/jax/quantization/unquantized.py (hunk 16 above)
Comment (JiriesKaileh) [reply tpu_inference/layers/jax/quantization/unquantized.py]: Removed

File: tpu_inference/models/jax/gemma4_mm.py
Comment (lk-chen) [inline tpu_inference/models/jax/gemma4_mm.py:1]: I'm not sure there should be any change in gemma4*.py

File: tpu_inference/layers/jax/quantization/unquantized.py
Comment (lk-chen) [inline tpu_inference/layers/jax/quantization/unquantized.py]: nit, add comment for shard_id

File: tpu_inference/layers/common/utils.py (hunk 13 above)
Comment (lk-chen) [reply tpu_inference/layers/common/utils.py]: still in latest revision

File: tpu_inference/models/jax/gemma4_mm.py
Comment (JiriesKaileh) [reply tpu_inference/models/jax/gemma4_mm.py:1]: Removed

File: tpu_inference/layers/jax/quantization/unquantized.py
Comment (JiriesKaileh) [reply tpu_inference/layers/jax/quantization/unquantized.py]: Added

**Review by lk-chen — APPROVED** [review]
